### PR TITLE
Move ids to type family

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR moves the `RawId` type into the type family. It is presented as `TF::DefId`, mirroring the name that rustc uses. The goal here is to allow rustc to map this to `DefId`.

For now, I kept the various newtypes (e.g., `StructId` and so forth) exactly as they are. At some point we are going to move to move things like `TypeName` so that we have an "interned" `TypeName`, but I'm not 100% sure about the various newtypes, they may want to stay like this, or else move into the `TypeFamily`, or maybe we just get rid of them (but I think it's nice to have the distinctions between various kinds of ids).

I think a good rule though would be that `TypeName` should be opaque to most of the chalk logic (apart perhaps from the clause generation logic, which would interact with it through methods on `RustIrDatabase`), which implies that we may want to extract some of its variants out into `TypeName`.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@detrumi I think you were poking at making a change like this, maybe you'd care to review?

File: chalk-solve/src/coherence.rs [hunk 1]
```diff
@@ -41,19 +41,25 @@ impl std::error::Error for CoherenceError {}
...
 #[derive(Clone, Debug, Default, PartialEq, Eq)]
-pub struct SpecializationPriorities {
-    map: BTreeMap<ImplId, SpecializationPriority>,
+pub struct SpecializationPriorities<TF: TypeFamily> {
+    map: BTreeMap<ImplId<TF>, SpecializationPriority>,
 }
 
-impl SpecializationPriorities {
+impl<TF: TypeFamily> SpecializationPriorities<TF> {
+    pub fn new() -> Self {
```
Comment (jackh726) [inline chalk-solve/src/coherence.rs]: If `SpecializationPriorities` derives `Default`, why is this needed?

Comment (nikomatsakis) [reply chalk-solve/src/coherence.rs]: True, we don't need this, we can write `::default`...

Comment (nikomatsakis) [reply chalk-solve/src/coherence.rs]: Ah, I remember. The problem is that `TF` doesn't necessarily implement `Default`. The fix would be either:

* to manually implement `Default` for `SpecializationPriorities<TF>` (for any `TF`, not just `TF: Default`)
* to require that all `TypeFamily: Default`, as we did with `Clone` and various other traits

The former is more typing that `new` which I think is why I didn't do it. The latter is actually a fine strategy so long as type families remain "dummy" types that just pull together a bunch of other logic. But I think that eventually we will want a `TypeFamily` to actually *be* an arena, in which case it may not want to implement `Default`. (i.e., in rustc, I expect a type family to be the `tcx`). That said, we could also make it so that you have `TF::Context` to represent the arena value -- but I think that may be more annoying in the end. I thought it'd be useful if we actually make a `TF` be the context itself, since that will help with type inference.

(I contemplated renaming `TF` to `TCX`... but I'm not sure we plan to keep that name in the compiler.)

**Comment by nikomatsakis** [comment]
I'm going to go ahead and merge this as is for now. Seems like the question of `Default` is not a major problem.
